KAFKA-3664: Commit offset of unsubscribed partitions of the new consumer on a subscription change

vahidhashemian opened the pull request:
When users are using group management, if they call consumer.subscribe() or consumer.unsubscribe() to change the subscription, the removed subscriptions will be immediately removed and their offset will not be committed.

This pull request fixes this issue by performing a commitAsync() in subscribe() and unsubscribe() methods to trigger an offset commit.

Special thanks to @hachikuji and @becketqin for helping out and sharing their feedback along the way.

vahidhashemian commented:
@becketqin @hachikuji I tried to follow @becketqin's design to implement this PR. A quick test on my side shows that the offsets are being committed now. Would love to have your feedback as I may have missed something here.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -280,6 +280,7 @@ protected void onJoinPrepare(int generation, String memberId) {
         log.info("Revoking previously assigned partitions {} for group {}", subscriptions.assignedPartitions(), groupId);
         try {
             Set<TopicPartition> revoked = new HashSet<>(subscriptions.assignedPartitions());
+            revoked.addAll(subscriptions.pendingUnsubscribePartitions().keySet());
```
It seems our latest agreement is that the partitions in pending unsubscribe state won't be passed to the rebalance listener. So should we leave this logic unchanged?

vahidhashemian replied:
Sorry I missed that. Fixed in the updated PR.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -102,6 +106,8 @@ public void subscribe(Collection<String> topics, ConsumerRebalanceListener liste
 
     public void changeSubscription(Collection<String> topicsToSubscribe) {
         if (!this.subscription.equals(new HashSet<>(topicsToSubscribe))) {
+            for (Map.Entry<TopicPartition, TopicPartitionState> entry: this.assignment.entrySet())
```
Maybe use `pendingUnsubscribe.putAll()`?

becketqin replied:
Actually maybe we can just add the unsubscribed partitions here.

vahidhashemian replied:
Sure. I hope I got it right in the updated PR.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -113,6 +119,10 @@ public void changeSubscription(Collection<String> topicsToSubscribe) {
                     it.remove();
             }
+            
+            // Remove from pending subscribe those partitions that remain assigned to the consumer
```
If we only add the unsubscribed partition in line 109 above, it looks we don't need line 123 to line 125 any more.

vahidhashemian replied:
Correct.

becketqin commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -260,6 +270,10 @@ public void seek(TopicPartition tp, long offset) {
         return this.assignment.keySet();
     }
 
+    public Map<TopicPartition, TopicPartitionState> pendingUnsubscribePartitions() {
```
Do we still need this method if we do not pass PendingUnsubscribe partitions to the rebalance listener?

vahidhashemian replied:
Right. Not needed anymore. Thanks for the quick feedback.

hachikuji commented:
@vahidhashemian Can you add some unit tests?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -53,6 +54,9 @@
     private final Set<String> subscription;
 
+    /* the list of topic partitions the user has unsubscribed from but not committed their offsets yet */
+    private final Map<TopicPartition, TopicPartitionState> pendingUnsubscribe;
```
Is it possible to use a flag in `TopicPartitionState` instead of a whole new collection?

vahidhashemian replied:
In the updated PR. I modified the implementation to avoid using a new collection.

vahidhashemian commented:
@hachikuji Thanks for reviewing the PR. It is updated as per your suggestion and now includes a unit test.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -283,10 +281,15 @@ public Long position(TopicPartition tp) {
+                allConsumed.put(tp, new OffsetAndMetadata(state.position));
+            // remove the unsubscribed topic partition from assigned partitions now that its offset is committed
+            if (state.pendingUnsubscribe)
+                this.assignment.remove(tp);
```
This mutation is a little surprising. I wonder if we should make removal explicit with a separate method (e.g. `clearPendingUnsubscribe()`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -363,12 +370,14 @@ public ConsumerRebalanceListener listener() {
         private OffsetAndMetadata committed;  // last committed position
         private boolean paused;  // whether this partition has been paused by the user
         private OffsetResetStrategy resetStrategy;  // the strategy to use if the offset needs resetting
+        private boolean pendingUnsubscribe; // whether this partition is pending unsubscribe from its current user
```
With this change, we may need to update some of the methods which read the `assignment` collection directly in `SubscriptionState` to filter out the methods. For example, don't we need to change `SubscriptionState.assignedPartitions()` and `SubscriptionState.pausedPartitions()`? Maybe we need some unit tests for `KafkaConsumer` to verify that unsubscribed partitions are not returned in `KafkaConsumer.assignment()`.

hachikuji replied:
One more note, does it make sense to just name this variable `unsubscribed` instead of `pendingUnsubscribe`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -102,17 +101,16 @@ public void subscribe(Collection<String> topics, ConsumerRebalanceListener liste
 
     public void changeSubscription(Collection<String> topicsToSubscribe) {
         if (!this.subscription.equals(new HashSet<>(topicsToSubscribe))) {
+            for (Map.Entry<TopicPartition, TopicPartitionState> entry: this.assignment.entrySet()) {
```
Was there a reason to move this block ahead in this function?

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/SubscriptionStateTest.java:
```diff
@@ -202,6 +203,20 @@ public void unsubscription() {
     }
 
+    @Test
+    public void pendingUnsubscribeAndOffsetCommit() {
```
This is a good start, but I think we need some additional assertions to validate that the behavior of other methods such as `assignment()` takes into account the unsubscribed partitions.

hachikuji commented:
@vahidhashemian Thanks for the patch. I think there's some details we should sort out as far as how this patch affects commit semantics. For example, if the user changes their subscription and immediately calls `commitSync()`, do we expect that offsets for the unsubscribed partitions should be committed? Or do we expect this patch to only affect autocommits? 

I'm also wondering about an alternative implementation which avoids the additional complexity in `SubscriptionState` by sending an asynchronous commit in `KafkaConsumer.subscribe()` prior to changing the subscription when autocommit is enabled. Maybe that would be a simpler if we only expect this patch to affect autocommit behavior?

@becketqin Thoughts?

vahidhashemian commented:
@hachikuji Thanks for sharing your thoughts and your feedback on the PR. I'll wait until a decision is made on which implementation option we should go with. I am wondering what you see as the issue with committing the unsubscribed offsets even when the autocommit is disabled. I am likely overlooking the details, but to me whether we are autocommitting or not we would want to properly keep track of whatever is consumed and so their offsets (former and current subscriptions alike) should be committed every time.

hachikuji commented:
@vahidhashemian As long as the expectation is clear to the user, I think either way can be OK, but I find it a little bit weird that we would commit offsets for partitions which are no longer contained in the user's assignment. We'll have to be very careful in the documentation to make sure this expectation is clear. I would slightly prefer keeping the assignment consistent with the offset commits so that user can always tell exactly which offsets will be committed in a given call to the commit API. 

This is why I suggested the change to use an async commit in `subscribe()` when autocommit is enabled. It maintains the current invariants as far as consistency between the assignment and offset commits, and it addresses the main problem from this JIRA, which is basically that users will see duplicates when they change subscriptions with autocommit enabled. For users who are using manual commit, they will have to call `commitSync()` prior to changing their subscription, but that seems reasonable to me.

vahidhashemian commented:
@hachikuji Thanks for your detailed explanation. They way I understand your alternate approach (and correct me if I'm wrong please) is that the [`subscribe()` method](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java#L829) would be changed to something like this:

```
    @Override
    public void subscribe(Collection<String> topics) {
        if (!subscriptions.assignedPartitions().isEmpty())
            commitAsync(subscriptions.allConsumed(), null);
        subscribe(topics, new NoOpConsumerRebalanceListener());
    }

```

Except that the auto commit status has to be checked there too (I haven't figured out yet how the consumer can see the auto commit status, which is in the coordinator).

hachikuji commented:
@vahidhashemian Yeah, that's what I was thinking, though the check in `subscribe()` would probably just look like this:

``` java
if (autoCommitEnabled)
  commitAsync();
```

The `autoCommitEnabled` flag comes from configuration.

vahidhashemian commented:
Thanks @hachikuji. The reason I added that check in subscribe() is it seems during the first subscribe() call when the consumer is just starting I get this exception:

```
ERROR Offset commit failed. (org.apache.kafka.clients.consumer.internals.ConsumerCoordinator:515)
org.apache.kafka.clients.consumer.RetriableCommitFailedException: Commit offsets failed with retriable exception. You should retry committing offsets.
Caused by: org.apache.kafka.common.errors.GroupCoordinatorNotAvailableException: The group coordinator is not available.
```

But if we somehow exclude that one, things seem to work as expected.

hachikuji commented:
@vahidhashemian Ah, I see. It looks like we're missing a check in `ConsumerCoordinator.commitOffsetsAsync()` for an empty collection. We have the check in `commitOffsetsSync()`, so perhaps we just need to add it.

@becketqin What do you think about this approach?

becketqin commented:
@hachikuji I think use `commitAsync()` is a better approach.

I am thinking about a corner case and not sure if it is possible. The case I am thinking is the following:
1. User called subscribe() and changed the subscription, an OffsetCommitRequest was put into the queue and not sent out yet.
2. When user call poll(), a rebalance finished before the OffsetCommitRequest in (1) is sent out. Now the OffsetCommitRequest will fail because it contains the partition that the consumer does not own anymore.

Could this happen?

hachikuji commented:
@becketqin That case may be possible, but I think it might be OK. If a rebalance is already underway and the consumer has sent its JoinGroup when `subscribe()` is called (which can actually happen if `wakeup()` is used in the middle of a rebalance), then any commit is already guaranteed to fail since it won't be handled by the coordinator until the JoinGroup has returned. Alternatively, if a rebalance has begun and the consumer has not yet sent its JoinGroup, then the network layer should guarantee that the commit is sent first. Does that seem right?

becketqin commented:
@hachikuji Agreed. I think it is working as design like you said. And it probably does not have a big impact either. I checked the code and it looks the entire rebalance process will not be interrupted unless user call `wakeup()`, so we can probably just do `commitAsync()`.

vahidhashemian committed:
KAFKA-3664: Commit offset of unsubscribed partitions of the new consumer on a subscription change

When users are using group management, if they call consumer.subscribe() or consumer.unsubscribe() to change the subscription, the removed subscriptions are immediately removed and their offset are not committed.
This pull request fixes this issue. It also cleans up the KafkaConsumerTest code.

vahidhashemian commented:
@hachikuji @becketqin I am trying to write unit tests for the updated approach, but am having some issues. It would be great if you could take a look, as you might spot the problem right away. The unit test is [here](https://github.com/apache/kafka/compare/trunk...vahidhashemian:KAFKA-3664-ALT2?expand=1#diff-444ce0cab1dc8ec666294c4eafc9baefR485) and the issue I am having is [this last line](https://github.com/apache/kafka/compare/trunk...vahidhashemian:KAFKA-3664-ALT2?expand=1#diff-444ce0cab1dc8ec666294c4eafc9baefR560) does not run properly and the test hangs there. I am basically trying to leverage the existing unit tests and change consumer subscription and make sure the last read offset is committed. I hope I am not far off. Thanks.

becketqin commented:
@vahidhashemian Since the commit was async, it looks we should do a `poll()` after changing the subscription. But even with the current code, hanging on line 560 still seems unexpected. Usually I would enable logging to see what is going on. @hachikuji Any idea?

vahidhashemian commented:
@becketqin Thanks for the suggestion. I enabled logging but did not see anything suspicious. If I use `consumer.poll(0)` it hangs forever in [this while loop](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java#L259). If I use `consumer.committed(partition)` it hangs forever in [this other while loop](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java#L162).

And the changes I've made in the branch mentioned above seem to be irrelevant. The updated unit test seems to fail with the current trunk too.

hachikuji commented:
@vahidhashemian It looks like you haven't mocked the offset commit response. This test uses `MockClient` which requires you to manually setup request expectations and the responses.

vahidhashemian commented:
@hachikuji Thank you for the pointer. I have not been able to find any documentation on how to use `MockClient`. So I'm trying to make sense of it by looking at the existing code.

hachikuji commented:
@vahidhashemian There probably isn't any documentation, but the usage should hopefully be clear enough based on other test cases. The consumer tests are a little tricky because there are multiple sockets which we're trying to mock without needing to depend on the relative order of the calls (since it doesn't matter). The offset commit should come from the coordinator connection, so if you follow the template of the other tests, you should just need to add something like this:

``` java
Struct offsetCommitResponse = buildOffsetCommitResponse();
client.prepareResponseFrom(offsetCommitResponse, coordinator);
```

And you can look in `ConsumerCoordinatorTest` to figure out how to build the offset commit response.

vahidhashemian commented:
@hachikuji Thanks again for your help. I looked at `ConsumerCoordinatorTest` and it seems to me that `OffsetCommitResponse` does not deal with actual offset value, but rather with committing the offset and making sure the commit is successful. Is it possible that `OffsetFetchResponse` is what's needed here? I actually tried `OffsetFetchResponse` in the unit test [here](https://github.com/vahidhashemian/kafka/blob/KAFKA-3664-ALT2/clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java#L476) without an error; but, even if the answer to my question above is "yes", I'm thinking I'm not doing everything right there. The reason is I am making up the offset fetch result, while I actually want to check what the offset value of `partition1` is as a result of running the unit test and changing `consumer1`'s subscription.

hachikuji commented:
@vahidhashemian You may need both. You definitely need the OffsetCommitResponse to avoid retrying the commit forever, and you may need the `OffsetFetchResponse` as well since the current behavior may be to send an offset fetch after every successful commit.

vahidhashemian commented:
@hachikuji In the meantime I'm trying to run a successful build, I managed to add two unit tests to verify the impact of the change when auto commit is enabled or disabled. Hopefully they are close to what you expected.

vahidhashemian commented:
@hachikuji @becketqin I would appreciate your feedback on the latest update (which has two unit tests for when the auto-commit is enabled or disabled). Thanks.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -519,6 +519,7 @@
     private final long retryBackoffMs;
     private final long requestTimeoutMs;
     private boolean closed = false;
+    private boolean autoCommitEnabled;
```
Seems like this could be made `final`.

vahidhashemian replied:
Nice catch. Will update.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -795,6 +799,8 @@ private KafkaConsumer(ConsumerConfig config,
     public void subscribe(Collection<String> topics, ConsumerRebalanceListener listener) {
         acquire();
         try {
+            if (this.autoCommitEnabled)
```
The call to `unsubscribe()` below already makes a call to `commitAsync`, so maybe it makes sense to move this to the `else` branch?

vahidhashemian replied:
That's true, will update this too.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -202,8 +204,8 @@ public void verifyHeartbeatSent() {
-        final KafkaConsumer<String, String> consumer = newConsumer(time, client, metadata, assignor,
-                sessionTimeoutMs, heartbeatIntervalMs, autoCommitIntervalMs);
+        final KafkaConsumer<String, String> consumer = newConsumer(time, client, groupId, metadata, assignor,
+                sessionTimeoutMs, heartbeatIntervalMs, Boolean.TRUE, autoCommitIntervalMs);
```
nitpick: it's a little weird to use `Boolean.TRUE` instead of just `true`.

vahidhashemian replied:
No problem, will modify as you suggested.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -469,6 +471,204 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+    }
+
+    @Test
+    public void testUnsubscribedPartitionOffsetCommitWithoutAutoCommit() {
```
It's not really clear to me why we need a test case with two consumers.

vahidhashemian replied:
I had initially reused the previous unit test and just modified the auto-commit property to `false` and the expected committed offset to `0`. However, the unit test raises an exception at the `consumer.unsubscribe()` line. Here's the exception with this approach:

```
org.apache.kafka.common.protocol.types.SchemaException: No such field: topics
    at org.apache.kafka.common.protocol.types.Struct.get(Struct.java:82)
    at org.apache.kafka.common.protocol.types.Struct.getArray(Struct.java:140)
    at org.apache.kafka.common.requests.OffsetCommitRequest.<init>(OffsetCommitRequest.java:198)
    at org.apache.kafka.clients.consumer.KafkaConsumerTest$11.matches(KafkaConsumerTest.java:627)
    at org.apache.kafka.clients.MockClient.send(MockClient.java:135)
    at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.trySend(ConsumerNetworkClient.java:350)
    at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:219)
    at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:178)
    at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.sendLeaveGroupRequest(AbstractCoordinator.java:589)
    at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.maybeLeaveGroup(AbstractCoordinator.java:566)
    at org.apache.kafka.clients.consumer.KafkaConsumer.unsubscribe(KafkaConsumer.java:880)
    at org.apache.kafka.clients.consumer.KafkaConsumerTest.testUnsubscribedPartitionOffsetCommitWithAutoCommitDisabled(KafkaConsumerTest.java:637)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:497)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
    at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
    at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:675)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
```

I could not find a better way around it, except using a second consumer that actually commits an offset for the topic partition. If there is a way I can avoid the exception with a single consumer please let me know.

hachikuji replied:
I see. What was the case you were trying to test here? Autocommit when unsubscribe() is used after assign()?

vahidhashemian replied:
This is the scenario I am testing here:
1. `consumer1` subscribes to `topic` (with only one partition) and has auto-commit enabled.
2. `consumer1` reads 5 records from the single topic partition from offset 0.
3. `consumer2` subscribes to the same `topic` with auto-commit disabled.
4. `consumer2` reads 5 records from the same partition starting from offset 5.
5. `consumer2` unsubscribes from `topic`.
6. `consumer1` unsubscribes from `topic`.
7. The last committed offset of the partition should be 5 (since `consumer2` does not commit offsets in `unsubscribe`).

Note that `consumer1` and `consumer2` belong to different consumer groups.

vahidhashemian replied:
I believe if we swap steps 5 and 6 above it would become more straightforward. `consumer1` unsubscribes committing the latest offset (5), and then `consumer2` unsubscribes not committing the latest offset (10).

hachikuji replied:
I'm not sure this makes sense. The offsets for each group are isolated, so `consumer2` would actually start from position 0. I think a better test case would be the following:
1. Start a single consumer with autocommit disabled.
2. Read 5 records.
3. Call unsubscribe().
4. Verify that no offset commit request was sent.

To be honest, this might be overkill, but I wouldn't complain if it was present.

vahidhashemian replied:
Ah, you are right. I'll try to make it simpler like the one you suggested.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -469,6 +471,204 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
     }
 
+    @Test
+    public void testUnsubscribedPartitionOffsetCommitWithAutoCommit() {
```
Do we have a test case which covers autocommit on `subscribe` as well?

vahidhashemian replied:
No, I'll try to create one with `subscribe`.

hachikuji commented:
@vahidhashemian Sorry for the delay. Left a few comments.

vahidhashemian commented:
Thanks @hachikuji for the feedback. Will respond shortly.

vahidhashemian commented:
@hachikuji I updated the pull request. Hopefully I was able to address your comments. Thanks in advance for taking another look.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -519,6 +519,7 @@
     private final long retryBackoffMs;
     private final long requestTimeoutMs;
     private boolean closed = false;
+    private final boolean autoCommitEnabled;
```
super nitpick: can we group all the final variables together?

vahidhashemian replied:
Sure, will swap with the previous line.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -469,6 +472,333 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+    
+        final KafkaConsumer<String, String> consumer = newConsumer(time, client, metadata, assignor,
+                sessionTimeoutMs, heartbeatIntervalMs, true, autoCommitIntervalMs);
+        consumer.subscribe(Arrays.asList(topic1, topic2), new ConsumerRebalanceListener() {
```
Minor: it seems like this test would be a little simpler if you start subscribed to topic1 and then switch to topic2.

vahidhashemian replied:
I'm running into an issue with this change. After the update and at the line where we call `subscribe()` a second time to change subscription from `topic1` to `topic2` the test raises an `Offset commit failed` error with a NullPointerException at [this line](https://github.com/vahidhashemian/kafka/blob/cdb7b7ba18070e53fe10616441fd90bab245a800/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L543). Any idea why?

vahidhashemian replied:
I'll change this too to make it simpler.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -469,6 +472,333 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+    
+        final KafkaConsumer<String, String> consumer = newConsumer(time, client, metadata, assignor,
+                sessionTimeoutMs, heartbeatIntervalMs, false, autoCommitIntervalMs);
+        consumer.subscribe(Arrays.asList(topic1, topic2), new ConsumerRebalanceListener() {
```
Since we add the same rebalance listener when we call subscribe() in all of these tests, maybe we should just turn it into a function?

vahidhashemian replied:
Yup. Makes sense.

hachikuji replied:
Much appreciated. I probably should have done this when I created the first of these test cases, but thanks for helping to clean things up.

vahidhashemian replied:
Happy to do it :)

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -469,6 +472,333 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+            }
+        });
+    
+        // lookup coordinator
```
Similarly, everything up to the fetch (i.e. coordinator lookup, join group, and sync group) are pretty much the same in all of these methods. Maybe we turn it into a function (e.g. `prepareRebalance`).

vahidhashemian commented:
@hachikuji I modified the unit tests and did some cleanups as you suggested, plus some extra cleanup that I thought would make unit tests more consistent. Thanks for your thorough reviews again.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -799,6 +803,8 @@ public void subscribe(Collection<String> topics, ConsumerRebalanceListener liste
                 // treat subscribing to empty topic list as the same as unsubscribing
                 this.unsubscribe();
             } else {
+                if (this.autoCommitEnabled)
```
Can you add a comment briefly saying why this commit is necessary?

vahidhashemian replied:
Will do in the next update.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -358,6 +358,9 @@ public void close() {
 
     public void commitOffsetsAsync(final Map<TopicPartition, OffsetAndMetadata> offsets, OffsetCommitCallback callback) {
+        if (offsets.isEmpty())
+            return;
```
Should have caught this before, but we should make sure the callback gets invoked as well.

vahidhashemian replied:
Right, adding it.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java:
```diff
@@ -421,8 +479,152 @@ public void testWakeupWithFetchDataAvailable() {
+        consumer.poll(0);
+
+        // respond to the outstanding fetch so that we have data available on the next poll
+        client.respondFrom(fetchResponse(partition1, 0, 5), node);
```
nitpick: seems like the offsets don't actually matter in this test (and the similar one for unsubscribe(). Maybe we could leave fetching out of the test?

vahidhashemian replied:
OK. At first I thought a fetch would make the use case more clear; but you're right. We just want to make sure no offset commit request is being sent. I'll remove them.

hachikuji commented:
@vahidhashemian This mostly looks good, but left a few minor comments.

vahidhashemian commented:
@hachikuji Thank you for the recent feedback. I updated the PR based on your suggestion.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -799,6 +803,9 @@ public void subscribe(Collection<String> topics, ConsumerRebalanceListener liste
                 this.unsubscribe();
             } else {
+                if (this.autoCommitEnabled)
+                    // make sure the offsets of topic partitions the consumer is unsubscribing from are committed
```
Maybe add "since these partitions will not be included in the revocation callback of the next rebalance"?

hachikuji commented:
@vahidhashemian Thanks for the iterations. LGTM.

vahidhashemian commented:
@hachikuji Thanks again for your time reviewing the PR multiple times.
cc @ewencp @ijuma

ewencp commented:
@vahidhashemian Sorry for the delay. This version LGTM given the current state of semantics for the consumer and how it modifies the `assignment` immediately upon subscription changes. I think having to handle this stuff both in the expected place (on revoke) and immediately in each subscription is a bit messy, but the right solution when preserving today's semantics.

I'm ready to commit this, but I think we should also consider whether those semantics should change before doing this (if we changed the semantics, some of this code would not be necessary). Here's my reasoning:
- I can't actually think of a pattern for application development that would rely on the current behavior where the assignment is filtered immediately.
- Current behavior seems error prone wrt resource cleanup since it isn't unreasonable for a user to expect symmetric calls to onPartitionsAssigned and onPartitionsRevoked.
- The behavior today is inconsistent between subscriptions that are a list of topics and regex subscriptions: when you set a regex subscription, the assignment is _not_ filtered immediately after the subscription change.
- Very few people probably rely on the intersection of functionality that would be affected: they would have to both pass in a ConsumerRebalanceListener, make subscription changes, and rely on the callback returning the filtered list.

I think it might be worth doing a quick KIP and trying to clean up the semantics if nobody objects. It would be an incompatible change which isn't great, but it is a semantic change (no API compatibility issues to worry about) and as mentioned above, I think in practice it likely doesn't affect many people directly now. Kafka Connect actually _could_ be affected by this if we get some config changes (like subscription) to pass through without requiring a full rebalance. It looks like we actually happen to track the necessary info such that this isn't a problem (because of some offset info we track in its own HashMap instead of relying on the assignment in the consumer), but minor changes to the code could easily break things.

hachikuji commented:
+1 on @ewencp's comment. I think this patch is worth keeping if we think the current (confusing) assignment semantics are worth preserving or if we don't think we have time to change the behavior before the next release. @vahidhashemian If you have time to take on a short KIP, I'd be happy to lend some help.

vahidhashemian commented:
@ewencp Thank you for reviewing this PR and the thorough explanation.
@hachikuji Sure, I'm up for taking on a KIP and definitely need some help.

ewencp commented:
@vahidhashemian Awesome :) It's more work to get the KIP through, but I really think it is the right solution.

I would just start with the [KIP template](https://cwiki.apache.org/confluence/display/KAFKA/KIP-Template), copy it, allocate a KIP number (on the [main KIP page](https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Improvement+Proposals)) and then just start filling in the text. I think the motivation will be clear given the discussion in this thread, but probably needs to be condensed to be very concise. The proposal is, I think, mostly straightforward. I think the compatibility section may need the most thorough discussion because we are talking about making a breaking semantic change in an interface that has been marked stable. In this section, we'll want to very clearly argue that the number of impacted users will be minimal to zero.

Both @hachikuji and I will be more than happy to review before sending out to the general kafka-dev list so we can try to iron out any kinks before circulating it more broadly.

vahidhashemian commented:
@ewencp @hachikuji I started a [KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-%23%23%3A+Revise+Partition+Assignment+Semantics+on+Subscription+Change) as you suggested, and filled the section based on some of the recent discussions on this PR. I'll publicized the KIP in the main KIP page once you've had a chance to take a look and advise whether I'm on the right track. By all means, please feel free to make changes to it as you review. Thanks.

ewencp commented:
@vahidhashemian What's the status on this? Do we just need to run a KIP-70 vote thread?

vahidhashemian commented:
@ewencp There was some positive feedback on the KIP. I could start a vote tomorrow if you think it's time to do that.

ewencp commented:
@vahidhashemian Yeah, definitely just push the KIP through with a vote if there weren't any objections or unanswered questions left.

vahidhashemian commented:
@ewencp Forgot to mention that [I had informed](http://mail-archives.apache.org/mod_mbox/kafka-dev/201607.mbox/%3cOFB2534DD4.69FBE5D2-ON00257FFD.005F3FC3-88257FFD.00600D8D@notes.na.collabserv.com%3e) Spark and Storm users about this potential change and except for one did not receive any feedback on those mailing lists. Not sure if we need to do anything further for that.

ewencp commented:
@vahidhashemian I think if we've made a best effort to solicit feedback from them, that's all we can really do.

vahidhashemian commented:
The fix for this defect was provided as part of a [PR](https://github.com/apache/kafka/pull/1726) for [KIP-70](https://cwiki.apache.org/confluence/display/KAFKA/KIP-70%3A+Revise+Partition+Assignment+Semantics+on+New+Consumer's+Subscription+Change) ([KAFKA-4033](https://issues.apache.org/jira/browse/KAFKA-4033)). Closing this PR.
